vllm/distributed/device_communicators/custom_all_reduce.py

```diff
@@ -174,6 +176,35 @@ def __init__(self,
                                        offsets, rank, self.full_nvlink)
         self.register_buffer(self.buffer)
 
+    @staticmethod
+    def create_shared_buffer(
+            size_in_bytes: int,
+            group: Optional[ProcessGroup] = None) -> List[int]:
+        lib = CudaRTLibrary()
+        pointer = lib.cudaMalloc(size_in_bytes)
+        handle = lib.cudaIpcGetMemHandle(pointer)
+        world_size = dist.get_world_size(group=group)
+        rank = dist.get_rank(group=group)
+        handles = [None] * world_size
+        dist.all_gather_object(handles, handle, group=group)
```

Comment:
Why do we need to use broadcast with `device=cpu` in `_gather_ipc_meta` but not here?

Reply:
when you use this function, the `group` argument should be `cpu_group` passed to custom allreduce object.

Reply:
see https://github.com/vllm-project/vllm/blob/40899855520eb9497606bdb2b1b4e619233e598a/vllm/distributed/parallel_state.py#L231-L236

Reply:
Why is `all_gather` fine here but not in `_gather_ipc_meta`?

Reply:
oh, that is because we met some issues with `all_gather` for tensors directly. here we are using `all_gather_object` , so it should be fine. see https://github.com/pytorch/pytorch/issues/126032 for the pytorch issue.

Reply:
I see. Someone refactored this to return a Tensor
https://github.com/vllm-project/vllm/pull/5047/files#diff-44d9d733ee604800cbce9858a9201db1044aeff2c940fa4a0521d0c9b6541b3eL137

A better way should be returning a string, if torch bindings doesn't support int8 directly.

Reply:
yeah string should be fine. #5047 aims to get rid of pybind11 so that we can release python version agnostic wheels.